diff --git a/Makefile b/Makefile
index 0fbaf7c..91a8db3 100644
--- a/Makefile
+++ b/Makefile
@@ -149,7 +149,11 @@ UPROGS=\
 	$U/_grind\
 	$U/_wc\
 	$U/_zombie\
-
+	$U/_sleep\
+	$U/_pingpong\
+	$U/_primes\
+	$U/_find\
+	$U/_xargs\
 
 ifeq ($(LAB),syscall)
 UPROGS += \
diff --git a/grade-lab-util b/grade-lab-util
index 10a2ba4..baa71c0 100755
--- a/grade-lab-util
+++ b/grade-lab-util
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin python
 
 import re
 from gradelib import *
